core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -667,4 +676,30 @@ public String toString() {
         }
         return "index {[" + index + "][" + type + "][" + id + "], source[" + sSource + "]}";
     }
+
+
+    /**
+     * Returns <code>true</code> if this request can have duplicates if {@link #isAutoGeneratedID()} is <code>true</code>.
+     * This is possible in the case the request is sent to the same shards more than once due to retries or network hickups
+     */
+    public boolean canHaveDuplicates() {
+        if (autoGeneratedID) {
+            return canHaveDuplicates;
+        }
+        return false;
```

Comment:
this is very tricky and has caused some bugs along the way. I added assertions and I want to keep it the way it is. lots of tests had it wrong